File: src/utils/normalizeItems.ts
export interface NormalizationResult<T> {
    items: T[],
    conflicts: Record<string, T[]>
}

export function normalizeItems<T>(
    items: T[],
    keySelector: (item: T) => string,
    merge: (key: string, item: T, other: T) => T | null,
): NormalizationResult<T> {
    const result: Map<string, T> = new Map()
    const conflicts: Map<string, T[]> = new Map()

    for (const item of items) {
        const key = keySelector(item)
        const existingItem = result.get(key)

        if (!existingItem) {
            result.set(key, item)
        } else {
            const mergedItem = merge(key, existingItem, item)

            if (mergedItem === null) {
                let existingConflict = conflicts.get(key)

                if (!existingConflict) {
                    existingConflict = []
                    conflicts.set(key, existingConflict)
                }

                existingConflict.push(item)
            } else {
                result.set(key, mergedItem)
            }
        }
    }

    return {
        items: Array.from(result.values()),
        conflicts: Object.fromEntries(conflicts.entries())
    }
}

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/util/normalizeItems.kt
package io.github.sgrishchenko.karakum.util

import js.array.ReadonlyArray
import js.objects.JsPlainObject
import js.objects.ReadonlyRecord
import js.objects.recordOf

@JsPlainObject
external interface NormalizationResult<T> {
    val items: ReadonlyArray<T>
    val conflicts: ReadonlyRecord<String, ReadonlyArray<T>>
}

fun <T> normalizeItems(
    items: ReadonlyArray<T>,
    keySelector: (item: T) -> String,
    merge: (key: String, item: T, other: T) -> T?,
): NormalizationResult<T> {
    val result = mutableMapOf<String, T>()
    val conflicts = mutableMapOf<String, MutableList<T>>()

    for (item in items) {
        val key = keySelector(item)
        val existingItem = result[key]

        if (existingItem == null) {
            result[key] = item
        } else {
            val mergedItem = merge(key, existingItem, item)

            if (mergedItem == null) {
                var existingConflict = conflicts[key]

                if (existingConflict == null) {
                    existingConflict = mutableListOf(existingItem)
                    conflicts[key] = existingConflict
                }

                existingConflict += item
            } else {
                result[key] = mergedItem
            }
        }
    }

    return NormalizationResult(
        items = result.values.toTypedArray(),
        conflicts = recordOf(
            pairs = conflicts
                .map { it.key to it.value.toTypedArray() }
                .toTypedArray()
        )
    )
}
